plugins/vbe/fu-vbe-device.c

```diff
@@ -0,0 +1,297 @@
+	gchar *fdt;
+	gint node;
+	GList *compat_list;
+	gchar *vbe_dir;
+} FuVbeDevicePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuVbeDevice, fu_vbe_device, FU_TYPE_DEVICE)
+#define GET_PRIVATE(o) (fu_vbe_device_get_instance_private(o))
+
+/**
+ * fu_vbe_device_get_method() - Get the VBE method for a device
+ *
+ * @self: Device to check
+ * Returns: method being used, e.g. "vbe-simple"
+ */
+const gchar *
+fu_vbe_device_get_method(FuVbeDevice *self)
+{
+	FuVbeDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), NULL);
+	return priv->vbe_method;
+}
+
+/**
+ * fu_vbe_device_get_fdt() - Get the FDT blob for a device
+ *
+ * This FDT is typically provided by the firmware and contains all the
+ * information relating to the device
+ *
+ * @self: Device to check
+ * Returns: pointer to FDT
+ */
+gpointer
+fu_vbe_device_get_fdt(FuVbeDevice *self)
+{
+	FuVbeDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), NULL);
+	return priv->fdt;
+}
+
+/**
+ * fu_vbe_device_get_node() - Get the FDT node for a device
+ *
+ * This node of the FDT contains all the information relating to the device
+ *
+ * @self: Device to check
+ * Returns: node offset
+ */
+gint
+fu_vbe_device_get_node(FuVbeDevice *self)
+{
+	FuVbeDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), -1);
+	return priv->node;
+}
+
+/**
+ * fu_vbe_device_get_compat_list() - Get the list of compatible strings for a device
+ *
+ * @self: Device to check
+ * Returns: list of compatible strings
+ */
+GList *
+fu_vbe_device_get_compat_list(FuVbeDevice *self)
+{
+	FuVbeDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), NULL);
+	return priv->compat_list;
+}
+
+/**
+ * fu_vbe_device_get_dir() - Get the director containing method information
+ *
+ * @self: Device to check
+ * Returns: method being used, e.g. "/var/local/lib/fwupd/vbe"
+ */
+const gchar *
+fu_vbe_device_get_dir(FuVbeDevice *self)
+{
+	FuVbeDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_val_if_fail(FU_IS_VBE_DEVICE(self), NULL);
+	return priv->vbe_dir;
+}
+
+static void
+fu_vbe_device_init(FuVbeDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self),
+			   FWUPD_DEVICE_FLAG_INTERNAL | FWUPD_DEVICE_FLAG_UPDATABLE |
+			       FWUPD_DEVICE_FLAG_NEEDS_REBOOT | FWUPD_DEVICE_FLAG_CAN_VERIFY |
+			       FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+
+	fu_device_add_protocol(FU_DEVICE(self), "org.vbe");
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_ENSURE_SEMVER);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_MD_SET_SIGNED);
+	fu_device_set_physical_id(FU_DEVICE(self), "vbe");
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_add_icon(FU_DEVICE(self), "computer");
+}
+
+static gboolean
+fu_vbe_device_probe(FuDevice *fud, GError **error)
+{
+	FuVbeDevicePrivate *priv;
+	FuVbeDevice *self = FU_VBE_DEVICE(fud);
```

Comment:
you shouldn't do this until after `g_return_val_if_fail`

Reply:
done